[color-scale] `editable` prop (MVP)

DmitrySharabin opened the pull request.

LeaVerou commented on src/color-scale/color-scale.js:
```diff
@@ -21,21 +21,98 @@ const Self = class ColorScale extends ColorElement {
+			return;
+		}
+
+		let colors = Object.entries(this.colors);
```
I’m missing some comments about what this block of code does overall

LeaVerou commented on src/color-scale/color-scale.js:
```diff
@@ -21,21 +21,98 @@ const Self = class ColorScale extends ColorElement {
+						textContent: "Add color",
+					});
+
+					addButton.addEventListener("click", evt => {
```
We should abstract the functionality into a method so that authors can trigger it programmatically (whether or not the `editable` class is there)

LeaVerou reviewed (changes requested):
See comments. Also I don't see anything about what keywords are used in `editable`

LeaVerou reviewed (changes requested):
Some (possibly non-exhaustive) comments:

 
- High level: I would have expected this to involve a lot more changes in `<color-swatch>`. Why is it trying to do everything in `<color-scale>`? A lot of this functionality is useful for individual swatches too. I'd expect mainly the list stuff to be in `<color-scale>`. Also, #107 seems to be causing issues here too (see below).

Delete button:
 - Remove button should be a part
 - I'd use "delete" rather than "remove" consistently
- The delete button should only appear on interaction, otherwise it adds too much visual noise
- The current styling is more "not allowed" rather than "delete"

Docs:
- Don't show the edge cases before the simple syntax. start from simple, then complex. I'd show the boolean `editable` first, and then the values to customize the behavior.

Other:
- Why are colors displayed twice in 
<img width="1006" alt="image" src="https://github.com/user-attachments/assets/6c0460f4-2df3-439a-8aca-2b86f030a1d3">
 - Good call on making interpolated colors not editable! Post-MVP, we could make the *number* of stops editable.
 - I would style the add button like an outlined swatch with a light gray border and a big + 
 - The only values I see for `editable` are `name` and `color`. How do I specify whether the list itself is editable and how? (add, delete, reorder, possibly with a `list` shorthand for all three)
- I wonder if a better UI for editing the color/name is a pencil icon next to it which makes it editable then. Or some way to choose.
 - Since UI is likely expected to reactively change the value of `edit` to toggle editing, I wonder if we should call it `edit` rather than `editable`. And we should be consistent across the library.

Food for thought:
- It would be good to start thinking how we can support the common need of modifying colors in a scale with a more elaborate UI, such as the one by our `<color-picker>` or even a third-party one, e.g. like 
<img width="317" alt="image" src="https://github.com/user-attachments/assets/dedb2f5e-6434-43a2-b405-64e09917ca83">

LeaVerou commented:
I know the change was my idea, but seeing it now, I think `editable` was a better name 🙈

Also, could we add an `editable` attribute on color swatches and use that rather than encoding color-swatch logic here?

DmitrySharabin commented:
> I know the change was my idea, but seeing it now, I think `editable` was a better name 🙈

No worries—it's an easy change. I see nothing wrong with trying things out and reverting to one of the previous options if the new ones fail.

> Also, could we add an `editable` attribute on color swatches and use that rather than encoding color-swatch logic here?

Sure. It was exactly the next step I was going to take. I just keep the things around so I don't lose progress.

DmitrySharabin commented:
I think the PR is ready for another iteration. Now, it does even more than before:

- `<color-swatch>` — https://deploy-preview-167--color-elements.netlify.app/src/color-swatch/#the-editable-attribute
- `<color-scale>` — https://deploy-preview-167--color-elements.netlify.app/src/color-scale/#the-editable-attribute

> * Why are colors displayed twice in
> 
> <img alt="image" width="1006" src="https://private-user-images.githubusercontent.com/175836/388689019-6c0460f4-2df3-439a-8aca-2b86f030a1d3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzM0NDEzOTUsIm5iZiI6MTczMzQ0MTA5NSwicGF0aCI6Ii8xNzU4MzYvMzg4Njg5MDE5LTZjMDQ2MGY0LTJkZjMtNDM5YS04YWNhLTJiODZmMDMwYTFkMy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMjA1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTIwNVQyMzI0NTVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0xZDk5MGM5MzgyZmZlZDU1MGE0MGQwMTU1MWI5MDkzMTVkMTRjOWRmZGViODE5OTdmY2MxZjM5ODE3OGJkOGMwJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9._t0TT5ZkyYjBTLqV2Nj4Fs_RQ6oQEYGGqJnh_Vf9_QQ">

The first are the color names, and the second are the colors themselves. Since we are editing colors here, the names are still visible. Anyway, I updated the example to be less confusing.

Still to implement: reordering colors, changing the number of "stops," enabling edit mode with a pencil button next to the color label, and changing colors with a popup. Probably, next iteration?

@LeaVerou, could you please give the PR another shot?

LeaVerou commented:
I have not yet properly reviewed it, just wanted to share some rendering glitches on Safari iOS:

![IMG_1916](https://github.com/user-attachments/assets/8ff9fdb0-a270-4fb2-8a19-6a52a224cef1)
![IMG_1917](https://github.com/user-attachments/assets/23c71440-29b7-4fe6-acbf-631a851cc540)
![IMG_1918](https://github.com/user-attachments/assets/88360d9f-e232-4abe-add5-1e4ccd93b007)

also, the + button needs and outline so that it looks a bit more visually similar to the swatches

can we make the name editing UI a bit less prominent than the color editing UI? When both are editable, there is currently nothing to guide the eye, and I think when both are editable, the color is edited a lot more

LeaVerou commented on src/color-scale/color-scale.css:
```diff
@@ -1,26 +1,85 @@
+}
+
+[part="add-button"],
+[part="delete-button"] {
```
I would suggest using "add button", then people can target them separately too (and you'd be able to simplify this to just `[part~="button"]

DmitrySharabin replied:
Excellent idea! Thanks.

LeaVerou commented on src/color-scale/color-scale.js:
```diff
@@ -17,31 +20,253 @@ const Self = class ColorScale extends ColorElement {
 			// Re-render swatches
+			// Only if nothing is being edited, otherwise the input would be lost
+			// or, e.g., "red" would be converted to "rgb(100%, 0%, 0%)" right after the typing is done
+			this.render();
```
I would suggest instead checking if it's focused (via `document.activeElement` IIRC). It's entirely possible that it's editable but nobody is interacting with it right now, in which case re-rendering would be fine, right?
Though re-rendering seems a very crude reaction to `computedColors` changing, when does this actually happen? Is it when the entire array if overwritten? In that case I guess it makes sense…

DmitrySharabin replied:
> I would suggest instead checking if it's focused (via `document.activeElement` IIRC). It's entirely possible that it's editable but nobody is interacting with it right now, in which case re-rendering would be fine, right?

You are absolutely right. Thank you for the idea. It works brilliantly. 

> Though re-rendering seems a very crude reaction to `computedColors` changing, when does this actually happen? Is it when the entire array if overwritten?

It does happen when `this.colors` is overwritten; you are right.

LeaVerou commented on src/color-scale/color-scale.js:
```diff
@@ -53,17 +278,56 @@ const Self = class ColorScale extends ColorElement {
+					if (!deleteButton) {
+						swatch.insertAdjacentHTML("beforeend", `
+							<button part="delete-button" title="Delete color">
+								<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
```
Where does this icon come from? Also you don't need the SVG namespace for inline SVG.

DmitrySharabin replied:
I took the icon from https://iconify.design/. I'll add a comment in the code.

> Also you don't need the SVG namespace for inline SVG.

Thanks. I'll fix it.

LeaVerou commented on src/color-scale/color-scale.js:
```diff
@@ -149,6 +416,53 @@ const Self = class ColorScale extends ColorElement {
+				from: true,
+			},
+		},
+		defaultColor: {
```
I think this should also accept a single color as a string or `Color` object (via an attribute too). Don't let the complex case get in the way of the simple one :)

DmitrySharabin replied:
Done!

LeaVerou commented on src/color-scale/color-scale.js:
```diff
@@ -17,31 +20,253 @@ const Self = class ColorScale extends ColorElement {
+		let {name: defaultName, color: defaultColor} = this.defaultColor?.() ?? {};
+
+		name ??= defaultName ?? "New color";
+		color ??= defaultColor ?? this.computedColors.at(-1)?.color ?? new Self.Color("#f06");
```
Nice try, but I'd use a middle gray as the ultimate fallback :)

DmitrySharabin replied:
> Nice try, but I'd use a middle gray (per OKLab, even if specified as a hex) as the ultimate fallback :)

I was thinking of an easter egg (sort of). 🙃 But yeah, a middle gray would be a good default.

> Also, more performant to specify a color space and array of coords than rely on parsing.

That's a good point. Thanks!

LeaVerou reviewed (changes requested):
See individual comments. But more big picture, this requires so much extra code, it makes me wonder if it would have been a better idea to have a separate component that is the editable version.

DmitrySharabin commented:
> See individual comments. But more big picture, this requires so much extra code, it makes me wonder if it would have been a better idea to have a separate component that is the editable version that inherits from this.

Makes perfect sense. Let me move in that direction. How would we call it? `<color-scale-editable>`?

DmitrySharabin commented on src/color-scale/color-scale.js:
```diff
@@ -17,31 +20,253 @@ const Self = class ColorScale extends ColorElement {
+	}
+
+	addColor (color, name) {
+		let {name: defaultName, color: defaultColor} = this.defaultColor?.() ?? {};
```
Honestly, I don't know why I made `defaultColor` a function. 🤦‍♂️ I'll make it a “simple” JS prop. Thanks for pointing out the issue.

DmitrySharabin commented:
I think I addressed all the issues you mentioned (please let me know if I'm not) and am ready to extract the code to two new components: `<color-swatch-input>` and `<color-scale-input>` (or whatever names we'll end up with).

LeaVerou commented:
Need to review this properly, but a couple things that jumped at me from a very cursory look:
- buttons need `cursor: pointer`. 
- the color-swatch docs still shows editable as slotting an input, rather than using the `editable` attribute. 
- this icon <img width="44" alt="image" src="https://github.com/user-attachments/assets/854badf6-4109-4b15-b189-3e4d2b26be2d" /> is a bit too complex. A pencil icon would suffice. And we should probably use FA icons :) 
- This is too visually complex: <img width="59" alt="image" src="https://github.com/user-attachments/assets/d713fe22-ae02-4bbb-ab82-bc77715f7944" /> I would do a simple icon with the right text color (which we know) and no background until it's hovered. Once hovered, _then_ it could have a single background (e.g. red with a white icon)
- I would rather avoid accent colors that may not work well with the host page if we don't need an accent color. I.e. I'd make this gray <img width="262" alt="image" src="https://github.com/user-attachments/assets/192182b3-a95e-43c3-b759-08cb5ebd3268" /> On hover it could preview what the default color is. It would also be nice to show "Add color" in the swatch name area when you hover.

LeaVerou reviewed (changes requested):
Still need to review the code, but left some comments about the design/docs.
